gzochi-server/doc/gzochi-dump.1
.TH GZOCHI-DUMP 1 "May 4, 2015"
.SH NAME
gzochi-dump \- Write gzochi databases to flat-text format
.SH SYNOPSIS
.B gzochi-dump
[-c <conf>] [-e <engine>] [-o <output_dir>] <app_name or data_dir>[:<db>]
.br
.B gzochi-dump
[-h | -v]
.SH DESCRIPTION
The gzochi-dump utility reads gzochi game application database files and 
writes them to standard output or a file using a portable flat-text format 
understood by the gzochi-load utility. This format is very similar to the 
format used by the dump and load tools for Berkeley DB, though the output of 
gzochi-dump can be read by the other gzochid tools no matter what storage 
engine the server is built against.
.PP
The argument to gzochi-dump is the name of an application or the path to a
directory containing an application's databases. If gzochi-dump detects that the
argument is not a directory, it assumes it must be an application name, and will
consult the server's gzochid.conf file to obtain the root of the server's
application data directory tree, which it uses to resolve the full path to the
application's database files. gzochi-dump may be used to dump any of the three 
gzochi game databases (\fBmeta\fR, \fBoids\fR, or \fBnames\fR) by specifying 
the database name as a suffix of the target. For example, the command:

.IP
gzochi-dump /var/gzochid/data/mygame:meta
.PP

will dump the contents of the \(dqmeta\(dq database to standard output. More 
usefully, however, gzochi-dump can dump all three databases to external files in
a single, consistent transaction, enabling online backups of all game data. For
example, the command:

.IP
gzochi-dump /var/gzochid/data/mygame
.PP

...will create the files \fBmeta.dump\fR, \fBoids.dump\fR, and \fBnames.dump\fR
in the current directory.

.SH OPTIONS
.IP \fB\-c,\ \-\-config\fR
Specify an alternate path to the gzochid.conf file.
.IP \fB\-e,\ \-\-engine\fR
Specify the name of the storage engine that created the target database. The
format of this argument is the same as for the \(dqstore.engine.dir\(dq
setting in gzochid.conf: the basename of the module file, minus the extension.
(E.g., bdb.)
.IP \fB\-o,\ \-\-output\fR
Specifies an output location for the data to be dumped. If dumping a single 
database, the argument names the complete path to a target output file, which
will be created or overwritten if it exists; if dumping all three game 
databases, the argument names the output directory, which must already exist,
into which the files \fBmeta.dump\fR, \fBoids.dump\fR, and \fBnames.dump\fR will
be written. If this argument is omitted, single databases will be dumped to 
standard output; the complete set of game database dump files will be writtento the current directory.
.IP \fB\-h,\ \-\-help\fR
Write usage information to standard error, and exit.
.IP \fB\-v,\ \-\-version\fR
Write version information to standard error, and exit.
.SH ENVIRONMENT
.TP
.B GZOCHID_CONF_LOCATION
If \fB$GZOCHID_CONF_LOCATION\fR is set before \fBgzochi-dump\fR is started, its
value is used as the full path to the server configuration file. This variable 
will be overriden by a command-line argument giving the location of the file, if
present.
.TP
.B GZOCHID_STORAGE_ENGINE_DIR
If \fB$GZOCHID_STORAGE_ENGINE_DIR\fR is set before \fBgzochi-dump\fR is started,
its value is used as the base path for resolving storage engine modules. This
value takes precedence over the \(dqstorage.engine.dir\(dq setting in the server
configuration file.

.SH AUTHOR
Julian Graham (joolean@undecidable.net)


NEWS
gzochi NEWS -- history of user-visible changes
Copyright (C) 2015 Julian Graham
See the end of copying conditions.

Send comments, questions, and bug reports to gzochi-devel@nongnu.org

Changes in 0.8 (since 0.7)

* Changes to the distribution

** Removed deprecated synchronous client (libgzochi)

libgzochi, the synchronous reference client implementation included with earlier
versions of the gzochi framework, has been removed from the distribution. The
client was deprecated in an earlier release, since the concurrency strategy
suggested by its API was undesirable. Application developers are encouraged to
use libgzochi-glib instead.

** Removed client support static library (libgzochi-client-common)

libgzochi-client-common, the static library that facilitated sharing of code
between libgzochi and libgzochi-glib, has been merged with libgzochi-glib and no
longer exists as an independent build target within the gzochi framework source
distribution.

** New example: External transactions

Client and server code for a new example application has been added to the 
source distribution under the "examples/" tree. This example demonstrates 
techniques for synchronizing a transaction executed within the gzochid container
with an external transaction-aware process, such as a relational database.


Changes in 0.7 (since 0.6)

* Changes to the distribution

** New reference client library implementation: (gzochi client)

Game developers who want to keep their entire application stack in Guile Scheme
(C server container notwithstanding) are in luck! The new (gzochi client)
library is a Scheme reference client implementation with an execution model that
can be integrated into a main loop or used stand-alone. The source distribution
for (gzochi client) is under the "gzochi-guile-client/" tree.


Changes in 0.6 (since 0.5)

* Compatibility

** Support for additional platforms

gzochi now builds on Mac OS X and (likely) other BSD variants.


Changes in 0.3 (since 0.2)

* Changes to the distribution

** New client support static library: libgzochi-client-common

Some of the trickier aspects of client communication with a gzochi server
instance, such as maintaining a buffer for partial message delivery, have been
factored out of libgzochi and into a static library, with a source distribution
under the "gzochi-client-common/" tree, that can be built into other client 
implementations. 

** New reference client library implementation: libgzochi-glib

libgzochi-client-common already has a consumer in the distribution:
libgzochi-glib, a new reference client implementation that integrates easily 
with the "main loop" functionality in the popular GNU GLib library to support
the development of single-threaded client applications. The source distribution
for libgzochi-glib is under the "gzochi-glib-client/" tree.

** Removed deprecated debugging client

mictlan, the primitive "debugging client" included with earlier versions of the
gzochi framework, has been removed from the distribution. Its functionality has
been superceded by the debugging capabilities of gzochid itself, which can be 
accessed using a telnet client.


Changes in 0.2 (since 0.1)

* Changes to the distribution

** New example: abermud

Client and server code for a new example game -- based on Alan Cox's 
AberMUD -- has been added to the source distribution under the "examples/"
tree. This example demonstrates techniques for maintaining a persistent world
in which state is shared between multiple players.


Copyright information:

Copyright (C) 2015 Julian Graham

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.


Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


gzochi-server/src/gzochi-dump.c
/* gzochi-dump.c: Utility for exporting data from game application databases
 * Copyright (C) 2015 Julian Graham
 *
 * gzochi is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <assert.h>
#include <config.h>
#include <errno.h>
#include <getopt.h>
#include <glib.h>
#include <locale.h>
#include <libintl.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "gzochid-storage.h"
#include "toollib.h"

#define _(String) gettext (String)

static void 
dump_header (FILE *to)
{
  fprintf (to, "VERSION=3\n");
  fprintf (to, "format=bytevalue\n");
  fprintf (to, "type=btree\n");
  fprintf (to, "HEADER=END\n");
}

static void
dump_key_value (char *k, size_t k_len, char *v, size_t v_len, FILE *to)
{
  int i = 0;

  assert (k != NULL);
  assert (v != NULL);
  
  fprintf (to, " ");
  for (; i < k_len; i++)
    fprintf (to, "%.2hhx", k[i]);
  fprintf (to, "\n ");
  for (i = 0; i < v_len; i++)
    fprintf (to, "%.2hhx", v[i]);
  fprintf (to, "\n");	  
}

static void
dump_store (gzochid_storage_engine_interface *iface, 
	    gzochid_storage_transaction *tx, gzochid_storage_store *store, 
	    FILE *to) 
{
  size_t key_len = 0;
  char *key = NULL;
  
  dump_header (to);
  key = iface->transaction_first_key (tx, store, &key_len);
  
  while (key != NULL)
    {
      char *old_key = key;
      size_t old_key_len = key_len;

      size_t value_len = 0;
      char *value = iface->transaction_get 
	(tx, store, key, key_len, &value_len);

      dump_key_value (key, key_len, value, value_len, to);
      key = iface->transaction_next_key (tx, store, key, old_key_len, &key_len);

      free (value);
      free (old_key);
    }
  
  fprintf (to, "DATA=END\n");
}

static void
dump_single_inner 
(gzochid_storage_engine_interface *iface, gzochid_storage_context *context, 
 gzochid_storage_transaction *tx, char *data_dir, char *db, FILE *to)
{
  char *db_path = g_strconcat (data_dir, "/", db, NULL);
  gzochid_storage_store *store = 
    gzochid_tool_open_store (iface, context, db_path);

  dump_store (iface, tx, store, to);

  iface->close_store (store);
  g_free (db_path);
}

static void
dump_single (gzochid_storage_engine_interface *iface, char *data_dir, char *db,
	     FILE *to)
{
  gzochid_storage_context *context = iface->initialize (data_dir);

  if (context == NULL)
    {
      g_critical ("Failed to initialize store in %s", data_dir);
      exit (EXIT_FAILURE);
    }
  else 
    {
      gzochid_storage_transaction *tx = iface->transaction_begin (context);
      dump_single_inner (iface, context, tx, data_dir, db, to);
      iface->transaction_rollback (tx);
      iface->close_context (context);
    }
}

static FILE *
open_dump_output_file (char *output_dir, char *file)
{
  FILE *output_file = NULL;
  char *filename = g_strconcat (output_dir, "/", file, NULL);
  
  output_file = fopen (filename, "w");
  if (output_file == NULL)
    {
      g_critical 
	("Failed to open file %s for writing: %s", filename, strerror (errno));
      exit (EXIT_FAILURE);
    }
  
  g_free (filename);
  return output_file;
}

static void
dump_all (gzochid_storage_engine_interface *iface, char *data_dir, 
	  char *output_dir)
{
  gzochid_storage_context *context = iface->initialize (data_dir);

  if (context == NULL)
    {
      g_critical ("Failed to initialize store in %s", data_dir);
      exit (EXIT_FAILURE);
    }
  else
    {
      FILE *meta_dump = open_dump_output_file (output_dir, "meta.dump");
      FILE *oids_dump = open_dump_output_file (output_dir, "oids.dump");
      FILE *names_dump = open_dump_output_file (output_dir, "names.dump");

      gzochid_storage_transaction *tx = iface->transaction_begin (context);

      dump_single_inner (iface, context, tx, data_dir, "meta", meta_dump);
      dump_single_inner (iface, context, tx, data_dir, "oids", oids_dump);
      dump_single_inner (iface, context, tx, data_dir, "names", names_dump);

      fclose (meta_dump);
      fclose (oids_dump);
      fclose (names_dump);

      iface->transaction_rollback (tx);
      iface->close_context (context);
    }
}

static const struct option longopts[] =
  {
    { "config", required_argument, NULL, 'c' },
    { "engine", required_argument, NULL, 'e' },
    { "output", required_argument, NULL, 'o' },
    { "help", no_argument, NULL, 'h' },
    { "version", no_argument, NULL, 'v' },
    { NULL, 0, NULL, 0 }
  };

static void
print_version (void)
{
  fprintf (stderr, "gzochi-dump (gzochi) %s\n", VERSION);

  fputs ("", stderr);
  fprintf (stderr, _("\
Copyright (C) %s Julian Graham\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n"),
          "2015");
}

static void
print_help (const char *program_name)
{
  fprintf (stderr, _("\
Usage: %s [-c <CONF>] [-e <ENGINE>] [-o <DIR>] <APP_NAME or DATA_DIR>[:<DB>]\n\
       %s [-h | -v]\n"), program_name, program_name);
  
  fputs ("", stderr);
  fputs (_("\
  -c, --config        full path to gzochid.conf\n\
  -e, --engine        the name of the storage engine in use, as per\n\
                      gzochid.conf (e.g., bdb)\n\
  -o, --output        the output directory, if dumping all databases; or\n\
                      the output file, if dumping a single database\n\
  -h, --help          display this help and exit\n\
  -v, --version       display version information and exit\n"), stderr);

  fputs ("", stderr);
  fprintf (stderr, _("\
Report bugs to: %s\n"), PACKAGE_BUGREPORT);
#ifdef PACKAGE_PACKAGER_BUG_REPORTS
  fprintf (stderr, _("Report %s bugs to: %s\n"), PACKAGE_PACKAGER,
          PACKAGE_PACKAGER_BUG_REPORTS);
#endif /* PACKAGE_PACKAGER_BUG_REPORTS */

#ifdef PACKAGE_URL
  fprintf (stderr, _("%s home page: <%s>\n"), PACKAGE_NAME, PACKAGE_URL);
#else
  fprintf (stderr, _("%s home page: <http://www.nongnu.org/%s/>\n"),
	   PACKAGE_NAME, PACKAGE);
#endif /* PACKAGE_URL */
}

int 
main (int argc, char *argv[])
{
  const char *program_name = argv[0];
  char *gzochid_conf_path = NULL;
  char *storage_engine_name = NULL;
  char *output = NULL;
  int optc = 0;
  
  setlocale (LC_ALL, "");
  
  while ((optc = getopt_long (argc, argv, "+ce:o:hv", longopts, NULL)) != -1)
    switch (optc)
      {
      case 'c':
	gzochid_conf_path = strdup (optarg);
	break;
      case 'e':
	storage_engine_name = strdup (optarg);
	break;
      case 'o':
	output = strdup (optarg);
	break;

      case 'v':
	print_version ();
	exit (EXIT_SUCCESS);
	break;
      case 'h':
	print_help (program_name);
	exit (EXIT_SUCCESS);
	break;

      default:
	print_help (program_name);
	exit (EXIT_FAILURE);
      }

  if (optind != argc - 1)
    {
      print_help (program_name);
      exit (EXIT_FAILURE);
    }
  else 
    {
      GHashTable *gzochid_conf = 
	gzochid_tool_load_game_config (gzochid_conf_path);

      char **targets = gzochid_tool_parse_targets (argv[optind]);
      char *data_dir = gzochid_tool_probe_data_dir 
	(gzochid_conf, targets[0], FALSE);
      char *db = targets[1];

      gzochid_storage_engine *engine = 
	gzochid_tool_probe_storage_engine (gzochid_conf, storage_engine_name); 

      if (db == NULL)
	{
	  if (output == NULL)
	    output = g_get_current_dir ();
	  dump_all (engine->interface, data_dir, output);
	  free (output);
	}
      else 
	{
	  FILE *output_file = NULL;

	  if (output == NULL)
	    output_file = stdout;
	  else 
	    {
	      if (g_file_test (output, G_FILE_TEST_IS_DIR))
		{
		  g_critical ("%s is a directory.", output);
		  exit (EXIT_FAILURE);
		}
	      else output_file = open_dump_output_file (output, db);

	      free (output);
	    }

	  dump_single (engine->interface, data_dir, db, output_file);
	}

      g_strfreev (targets);
    }

  return 0;
}
